Tyson/mfsim trans

tysolov opened the pull request.

tysolov committed:
Working on MFSim target.  DAG creation is complete. CFG creation is underway.

TODOs added for retooling how control transfers are found when SSA renaming is fixed.

Discovered that, AFAIK, instruction properties (time, volume, temp, etc) are discovered but not being stored.

MFSim target is correctly outputting for initial test case (pcr.bs), which contains a dispense, several heats, a detect and a repeat block.

Still need to test/implement translation for:
   -- while loops
   -- if/else blocks
   -- nested conditionals
   -- compound conditional expressions

tysolov committed:
Working on MFSim target.  DAG creation is complete. CFG creation is underway.

TODOs added for retooling how control transfers are found when SSA renaming is fixed.

Discovered that, AFAIK, instruction properties (time, volume, temp, etc) are discovered but not being stored.

MFSim target is correctly outputting for initial test case (pcr.bs), which contains a dispense, several heats, a detect and a repeat block.

Still need to test/implement translation for:
   -- while loops
   -- if/else blocks
   -- nested conditionals
   -- compound conditional expressions

tysolov committed:
Working on MFSim target. CFG creation nearly complete.

This commit:
 -- CFG edges for if/else blocks being generated
 -- Tweaks to how transfer_in/transfer_out nodes are generated to account for disposes [i.e. liveness of a var ends]
 -- bug: introduced that causes a thread join error (lock competition) when returning from main.py -- no idea how/where/why this is happening.

Notes on assay translation tested this commit:
   -- probabilistic_pcr.bs is translating correctly
   -- bug: pcr.bs is inserting 2 extraneous transfer_out nodes to DAG4 between the detect and the dipose

Next:
   -- hunt down bugs introduced here
   -- nested conditionals/loops
   -- test full flow from BS->MFSim->activations on common assays

tysolov committed:
Working on MFSim target. CFG creation nearly complete.

This commit:
 -- CFG edges for if/else blocks being generated
 -- Tweaks to how transfer_in/transfer_out nodes are generated to account for disposes [i.e. liveness of a var ends]
 -- bug: introduced that causes a thread join error (lock competition) when returning from main.py -- no idea how/where/why this is happening.

Notes on assay translation tested this commit:
   -- probabilistic_pcr.bs is translating correctly
   -- bug: pcr.bs is inserting 2 extraneous transfer_out nodes to DAG4 between the detect and the dipose

Next:
   -- hunt down bugs introduced here
   -- nested conditionals/loops
   -- test full flow from BS->MFSim->activations on common assays

tysolov committed:
bug fixes on MFSim targetting

tysolov committed:
bug fixes on MFSim targetting

tysolov committed:
Fixes bug for detect instruction

tysolov committed:
Fixes bug for detect instruction

tysolov committed:
Cleaning up

tysolov committed:
Cleaning up

tysolov committed:
updated grammar

tysolov committed:
Cleaning up

tysolov committed:
Cleaning up

tysolov committed:
Bug fix for MFSim target:

details: when a non-split droplet has multiple successors (this occurs anytime we detect a droplet, as detect does not 'consume' this droplet), we previously drew edges to all successors.  We now find the use(s) that does not rename the droplet (currently only a detect will cause this) and only draw edges to those uses.

This enforces the following semantic:  each droplet (in the generated MFSim DAGs) is defined and used at most once.

tysolov committed:
Bug fix for MFSim target:

details: when a non-split droplet has multiple successors (this occurs anytime we detect a droplet, as detect does not 'consume' this droplet), we previously drew edges to all successors.  We now find the use(s) that does not rename the droplet (currently only a detect will cause this) and only draw edges to those uses.

This enforces the following semantic:  each droplet (in the generated MFSim DAGs) is defined and used at most once.

lilott8 committed:
refactoring the symbols and its visitor.

lilott8 committed:
volume tracking in mix works.

lilott8 committed:
abstracted fluidic properties.  It almost works completely.

lilott8 committed:
Testing framework built out.  Mixes now work.

lilott8 committed:
Forced a test to check specific indices.

lilott8 committed:
forgot to add a few assertions.

lilott8 committed:
detection is working.

lilott8 committed:
heat instructions work + test cases.

lilott8 committed:
split is partially done.  It passes the two tests, but I have more to do to make sure it works.

lilott8 committed:
Renamed dispense tests.

tysolov committed:
Updating README

Commit:
Update README.md

tysolov committed:
Update README.md for quick+easy submodule updates.

tysolov committed:
Updated README for quick+easy submodule updating.

lilott8 committed:
Updated symbol table and test.  The symbol table visitor only deals
with building elementary symbols.

lilott8 committed:
all instructions except method calls are done, replete with test cases.

lilott8 committed:
math is done, methods are almost done.  Now onto conditionals.

lilott8 committed:
added a few more tests for functions.

lilott8 committed:
method chaining is working.

lilott8 committed:
added tests for numeric assignment and gradient.

lilott8 committed:
getting new IR structures in place.

lilott8 committed:
dispense and mix ir conversion is done.

lilott8 committed:
detect is working now.

lilott8 committed:
heat is passing its tests now.

lilott8 committed:
split is working.

lilott8 committed:
dispose is functional.

lilott8 committed:
store works.

lilott8 committed:
numbers and math are working.

lilott8 committed:
numbers and math are working.

lilott8 committed:
if statements work only with simple operations (no && or ||)

lilott8 committed:
Simple conditions are fully tested with if statements.

lilott8 committed:
while loops insert the appropriate jumps.

lilott8 committed:
if{while{}} graphs are now correct.

lilott8 committed:
if{while{}} graphs are now correct.

lilott8 committed:
functions now work correctly.

lilott8 committed:
functions now work correctly.

lilott8 committed:
Fixed a discrepancy with globalz in the ir visitor and symbol table.

lilott8 committed:
inkwell is almost functional with the new semantics.

lilott8 committed:
This almost works.  A mix leading into a split doesn't function yet.
This is because the operation doesn't know to look at the ancestor edges to group them all into one operation.

lilott8 committed:
This generates the dag correctly.  I need to fix the translation from dag to inkwell.

lilott8 committed:
got the graph working correctly.

lilott8 committed:
Fixed an issue with connections being built more than once.

lilott8 committed:
fixed the split module's port placement.

Commit:
Merge branch 'master' into tyson/mfsim_trans

lilott8 committed:
python was moved to a virtual_env.  Thus, requirements.txt has changed.

lilott8 committed:
phi node insertion is behaving.  But I need to actually insert the node(s).

lilott8 committed:
Got mix and dispense expansion working.

lilott8 committed:
Got mix and dispense expansion working.

lilott8 committed:
bug fix: input-ports were not using all ports.

lilott8 committed:
Got disposals working.

lilott8 committed:
Merge branch 'workflow' of github.com:lilott8/BioScript into workflow

lilott8 committed:
Heat is working.

lilott8 committed:
Detect is working.

lilott8 committed:
I pass all test cases.

lilott8 committed:
Re-enabled SSA

lilott8 committed:
Added a series of assays from old literature.

lilott8 committed:
Added an extra test for conditionals.

lilott8 committed:
a logging version of SSA is almost finished for _one_ case.

lilott8 committed:
Variable renaming is working correctly.
fixes #56
fixes #58
fixes #57

lilott8 committed:
Variable renaming is working correctly.
fixes #56
fixes #58
fixes #57

lilott8 committed:
Merge pull request #60 from lilott8/workflow

Workflow

lilott8 reviewed (changes requested):
This needs to be rebased and fixed.

lilott8 committed:
removed phi nodes that have no uses.

lilott8 committed:
Merge pull request #61 from lilott8/hotfixes

removed phi nodes that have no uses.

lilott8 committed:
removed an unneeded file.

lilott8 committed:
Merge pull request #52 from lilott8/update-submodule-info

Update submodule info

lilott8 committed:
Updated readme and requirements.

lilott8 committed:
Merge pull request #62 from lilott8/hotfixes

Hotfixes

lilott8 committed:
Updating the grammar

lilott8 committed:
Fixed some odd issues with a few changes to the grammar.

lilott8 committed:
Merge pull request #63 from lilott8/dev

Fixed some odd issues with a few changes to the grammar.

lilott8 committed:
added planarity test.

lilott8 committed:
Merge pull request #64 from lilott8/dev

added planarity test.

lilott8 committed:
Fixed a few tests, and changed the way graphs are drawn.

lilott8 committed:
Merge pull request #65 from lilott8/dev

Fixed a few tests, and changed the way graphs are drawn.

lilott8 committed:
renamed a few assays.

lilott8 committed:
Merge pull request #66 from lilott8/dev

renamed a few assays.

tysolov committed:
Repeat loop is used more often than while/if loops in our suite of benchmarks.

Recent updates broke repeat -> while translation; repeat now working again.

tysolov committed:
Constant assignment now tracking correctly in the symbol table
SSA translation with numerical operations functioning properly
Bug that arose when defining/using a numerical var in the same math operation squashed.

tysolov committed:
- If output directory does not exist, creates it
- SSA pass now updates def/use names at block level
- repeat->while IR transformation corrected
- MFSim target updated for adherence to new IR
 (NOTE: MFSim target does not currently support SIMD semantics; to support these, we will need to unroll semantic meaning during targeting for discrete droplets in a SIMD variable)

tysolov committed:
MFSim targeting updated to work for splits of size 2 only.  TODO: generalize split target for n droplets
Need to complete array indexing for all instructions when splits are present --- in general need to implement SIMD semantics

tysolov committed:
Bug squashes, test added for repeat->while translation

tysolov committed:
Moar repeat tests added

tysolov committed:
Merge remote-tracking branch 'origin/tyson/mfsim_trans' into tyson/mfsim_trans

# Conflicts:
#	.gitignore
#	compiler/data_structures/ir.py
#	compiler/passes/transforms/ssa.py
#	compiler/semantics/ir_visitor.py
#	compiler/semantics/symbol_visitor.py
#	compiler/targets/mfsim_target.py
#	resources/assays/pcr.bs
